### ISSUE TITLE
Finetuning with small dataset

### ISSUE THREAD

**Issue by LeDuySon** [ZhengPeng7/BiRefNet#24]
First of all, thank you for your great project. I want to ask you for some recommendation about finetuning with a small dataset (Around 400 images), my problem is main car segmentation (Only segment one car even if the image contains multiple cars, the main car is the biggest one and in the middle of the image)
- Which layers should i freeze?
- What learning rate should i start with?
- Because the car segmentation is pretty simple, should i turn off any loss components?
- Do you have any idea of how many images is pretty enough for training?
Due to the limit of resources so i can't try all these thing, your suggestion will be really meaningful to me.

**Comment by ZhengPeng7** [comment]
1. Model: Since you have only a little data, I suggest using a smaller model, e.g., choose the `swin_v1_tiny` as the backbone in `config.py` (remember to put the weights of backbone in the right place).
2. Freezing layers: I'm not sure whether freezing some layers can help the training. But if you want, you can turn on the `freeze_bb` option in `config.py` to easily freeze the layers of the backbone.
3. Loss: You can turn off the ssim loss in `config.py` since it benefits segmentation in fine regions, which is unnecessary in your case. In my experience, IoU loss converges much faster but decreases in accuracy. But if you want to see the results faster, you can leave it on only.
4. If you do not have extra data, you can split 40 images for validation.

If you have further questions, feel free to leave messages :)

**Comment by LeDuySon** [comment]
Thank you so much for your detailed answer. 

About the swin_v1_tiny, can we have a massive dataset training with this one? I found the massive training one is much better in general case compare to the one that trained on only one dataset. 

About the loss function, i think about it again because my input image can sometime be like this, kinda complex so i will need to try it myself, do you have any recommendation about where to hire gpu?

![image](https://github.com/ZhengPeng7/BiRefNet/assets/33374938/e9496d4a-ed16-4187-90c7-5b58316c076f)

**Comment by ZhengPeng7** [comment]
Thanks for your feedback! However, it takes a lot of time to do the massive training, even with swin-tiny, which I haven't started. I can only say I might spare my own and GPUs' time for it in the future.

If your cases are similar to the image above, I recommend using the default settings of losses in my project.

About renting GPUs, I personally recommend those on autodl, which is the cheapest platform I used. But if you are not in China (you know there are firewalls to block people from things like Google), I recommend finding some GPUs on vast.ai. BTW, if you want to use the default training setting (bs=2, bb=Swin-L), you need GPUs with more than 37G memory. If you want to train with swin-tiny, you can use GPUs with 24G memory. Batch size is better to be larger than 1 (I've tested full training with bs=1).

**Comment by LeDuySon** [comment]
Thank you! 
Have you tried to export this model to Onnx? I plan to deploy this model to triton inference server latter after the training so if you have not, maybe i will try to do it and get back to you

**Comment by ZhengPeng7** [comment]
Sorry, I haven't done this kind of thing. But if you encounter some problems while doing the deployment, which you think I may know about, feel free to leave messages here. Good luck!

**Comment by LeDuySon** [comment]
Thank you!

**Comment by LeDuySon** [comment]
Hi @ZhengPeng7 , i know this one is not related to this discussion but i can't load the BiRefNet_DIS_ep500-swin_v1_tiny anymore? Do you know why? I have changed the backbone in config to swin_v1_t but when loading the checkpoint, it just shows mismatched between many layers

**Comment by ZhengPeng7** [comment]
There were some differences between the previous codes and the descriptions in the paper in terms of model architecture. I made the modifications so that they are 100% correct with each other now.
I'll try to train a swin-tiny version in the massive training setting. I'll reply to you once it's done.

**Comment by ZhengPeng7** [comment]
Feel free to reopen it if you have any more question.

**Comment by LeDuySon** [comment]
@ZhengPeng7 I have finished my training on tiny version with around 1500 annotated images. Thank for your suggestions, i want to ask about some ideas to improve the current model. Current model have some problems attached in the images below
![image](https://github.com/user-attachments/assets/9367eef1-1a87-41ca-8fe2-ea0df792b7d0)
![image](https://github.com/user-attachments/assets/83f95116-884d-4783-ac27-a94616bfd573)
- The contour around the object have some weird artifact? Do you have any idea to fix this, i'm thinking is due to the mask resize when inference
- Some images have multiple cars but i only want to segment the main car (the car in the middle of the images) so sometime it will segment the car nearby too

Questions:
- Do you have any idea or reference to handle these cases?
- Why you only apply the flip augmentation to the dataset?
- Any suggestion on the hyperparameter setting for full finetuning the tiny version? The learning rate, loss setting

Thank you!

**Comment by ZhengPeng7** [comment]
Hi, @LeDuySon,
1. I'm not very sure about the reason but the one you provide makes sense. Besides, I don't know the sizes of

**Comment by IndieMinimalist** [comment]
> There were some differences between the previous codes and the descriptions in the paper in terms of model architecture. I made the modifications so that they are 100% correct with each other now. I'll try to train a swin-tiny version in the massive training setting. I'll reply to you once it's done.

Hi, @ZhengPeng7
Do you have the swin-tiny model in safetensors format that I can use for inference?

**Comment by LeDuySon** [comment]
Hi @ZhengPeng7 ,
Sorry for late reply and thank you for your suggestion!
1. Yes, my image is bigger compare to the model size (some images even 4k) so i think thats the reason. Do you have any idea to mitigate this problem?
2. I have tried with instance segmentation but the mask is not really good, maybe i can do both instance segmentation and your model but the inference time will be much higher. And i just checked you box guided segmentation, maybe i will have a look on it because the main car always in the center of the image
3. Do you have any suggestions for augmentation in my cases? I think i will try to add more augmentation and experiment with it
4. Thank you, i will try that

**Comment by cyy-1234** [comment]
Hi，Great creator, I would like to ask, when predicting, is it necessary to resize to 1024x1024 to get better results, because I noticed that the actual input image size, aspect ratio is not fixed when scaling is not etc. Will the segmentation accuracy be affected when scaling is performed?
